[clang-tidy] Added bugprone-unsequenced-global-accesses check

ConcreteCactus opened the pull request:
This checker attempts to cover the global variable case in rule [EXP-30](https://wiki.sei.cmu.edu/confluence/display/c/EXP30-C.+Do+not+depend+on+the+order+of+evaluation+for+side+effects) in the SEI CERT C and rule [EXP-50](https://wiki.sei.cmu.edu/confluence/display/cplusplus/EXP50-CPP.+Do+not+depend+on+the+order+of+evaluation+for+side+effects) in the SEI CERT C++ Coding Standards.

It recurses into functions in the same translation unit and can handle global struct and union members as well.

ConcreteCactus committed:
[clang-tidy] Added bugprone-unsequenced-global-accesses check

    This check attempts to detect unsequenced accesses to global
    variables. It recurses into function calls in the same translation unit,
    and can handle fields on global structs/unions.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/ConflictingGlobalAccesses.h:
```diff
@@ -0,0 +1,77 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_TOOLS_EXTRA_CLANG_TIDY_BUGPRONE_SIDEEFFECTBETWEENSEQUENCE\
+POINTSCHECK_H
+#define LLVM_CLANG_TOOLS_EXTRA_CLANG_TIDY_BUGPRONE_SIDEEFFECTBETWEENSEQUENCE\
+POINTSCHECK_H
```
wrong name

ConcreteCactus replied:
Done, I changed the header guard to the original name.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/ConflictingGlobalAccesses.h:
```diff
@@ -0,0 +1,77 @@
+/// For example: \code
+///
+/// int a = 0;
+/// int b = (a++) - a; // This is flagged by -Wunsequenced.
+///
+/// \endcode
+///
+/// However global variables allow for more complex scenarios that
+/// -Wunsequenced doesn't detect. E.g. \code
+///
+/// int globalVar = 0;
+///
+/// int incFun() {
+///   globalVar++;
+///   return globalVar;
+/// }
+///
+/// int main() {
+///   return globalVar + incFun(); // This is not detected by -Wunsequenced.
+/// }
+///
+/// \endcode
+///
+/// This checker attempts to detect such undefined behavior. It recurses into
+/// functions that are inside the same translation unit. It also attempts not to
+/// flag cases that are already covered by -Wunsequenced. Global unions and
+/// structs are also handled.
+///
+/// It's possible to enable handling mutable reference and pointer function
+/// parameters as writes using the HandleMutableFunctionParametersAsWrites
+/// option. For example: \code
+///
+/// void func(int& a);
+///
+/// int globalVar;
+/// func(globalVar); // <- this could be a write to globalVar.
+///
+/// \endcode
+///
+/// This option is disabled by default.
```
this should be in documentation file, not here

ConcreteCactus replied:
I put the comment into the documentation file. Please, let me know if you feel that more documentation is needed.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/ConflictingGlobalAccesses.cpp:
```diff
@@ -0,0 +1,748 @@
+  ast_matchers::internal::Matcher<Expr> GlobalAccessMatcher =
+      twoGlobalWritesBetweenSequencePoints(LangStd);
+
+  Finder->addMatcher(
+      stmt(traverse(TK_AsIs, expr(GlobalAccessMatcher).bind("gw"))), this);
+}
```
this matcher will be executed way to often.
twoGlobalWritesBetweenSequencePoints doesnt need to exist, match here call expr and binary operator directly.
what about overloaded operators ?

ConcreteCactus replied:
I think overloaded operators are deriving from call expressions, so those should get checked as well. But correct me if I'm wrong.

I'll refactor the logic into the check call.

ConcreteCactus replied:
I now match on binary operators and call expressions directly. I also added matchers to `InitListExpr`, `CXXConstructExpr`, and `ArraySubscriptExpr`, as those aren't always sequenced either.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/ConflictingGlobalAccesses.cpp:
```diff
@@ -0,0 +1,748 @@
+};
+
+/// GlobalRWVisitor (global read write visitor) does all the traversals.
+class GlobalRWVisitor : public RecursiveASTVisitor<GlobalRWVisitor> {
```
i'm not sure about this, look into check bugprone-unchecked-optional, it uses different framework, maybe it could be used here, or look into framework used by bugprone-unhanded-exception.
For me this "call" based framework could be extracted into some utility class, and made available for everyone, so dev could derive from it, implement some handlers/visitors and use, instead of re-implementing it. And this does not cover everything, like modify of global object from destructor of temporary variable.

ConcreteCactus replied:
Thanks, I'll look into how those checkers solve this.

ConcreteCactus replied:
I looked into the framework used by bugprone-unchecked-optional. I think using the data flow analysis framework has some benefits and some drawbacks. For example, using a more thorough framework could help eliminate false-positives that require branch evaluation, and there wouldn't be a need for this `GlobalRWVisitor`. However, the checker would probably get slower, and it would need to be rewritten substantially. Maybe even moved into clang-sa.
I would prefer using this `GlobalRWVisitor`, but if you still think it's better to do a data flow analysis, then I will rewrite the checker to use that.

By the way, I couldn't find a checker named bugprone-unhandled-exception in clang-tidy. Only one named `bugprone-unhandled-exception-at-new`. But I don't think you are referring to that.

vbvictor replied:
Exception check is called `bugprone-exception-escape` and the checker file is `ExceptionEscapeCheck.cpp`. It uses framework defined in `ExceptionAnalyzer.cpp`.

ConcreteCactus replied:
Looks like the `ExceptionAnalyzer` does its own recursion instead of relying to `RecursiveASTVisitor`. I think in my case its easier to rely on the visitor because this checker requires recursing into expressions which I find complicated.

In my last commit, I moved some of the functionality in `GlobalRWVisitor` to a new utility called `ExecutionVisitor`. This `ExecutionVisitor` contains the parts of the old `GlobalRWVisitor` that in my opinion could be useful in other checkers. That is, `ExecutionVisitor` traverses statements that could be executed while executing another statement. It will traverse function bodies of call expressions, initializers and bodies of constructors, and also destructors if they could be called.

PiotrZSL reviewed (commented):
Not sure about name, some simpler name would be better.
Missing documentation/release notes.

Add check need to be simplified.

carlosgalvezp commented:
If this is a CERT check, please add the corresponding aliases towards the CERT module.

vbvictor commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/conflicting-global-accesses.cpp:
```diff
@@ -0,0 +1,290 @@
+// RUN: %check_clang_tidy -std=c++11 -check-suffixes=,CPP11 %s bugprone-conflicting-global-accesses %t
```
Should this check work prior to `c++11`? I don't see any reasons why it need `c++11`.
We could remove this argument

ConcreteCactus replied:
The checker flags some C++11 specific errors that shouldn't be flagged in later versions of the standard. (e.g. bitshift operators had their order of evaluation defined in C++17) Hence the argument. The default value to `-std` is `C++11-or-later`, which is not quite what I need. But I agree that the checker runs are missing a general case, so I'll change the `C++17` ones to `C++17-or-later`.

vbvictor replied:
We can do `-std=c++11,c++14` for this case then

vbvictor commented on clang-tools-extra/clang-tidy/bugprone/ConflictingGlobalAccesses.cpp:
```diff
@@ -0,0 +1,748 @@
+  const LangStandard *LangStd =
+      &LangStandard::getLangStandardForKind(getLangOpts().LangStd);
+
+  ast_matchers::internal::Matcher<Expr> GlobalAccessMatcher =
```
nit: usually we use `auto` instead of writing explicit type for matchers

vbvictor commented on clang-tools-extra/clang-tidy/bugprone/ConflictingGlobalAccesses.cpp:
```diff
@@ -0,0 +1,748 @@
+}
+
+bool GlobalRWVisitor::VisitDeclRefExpr(DeclRefExpr *DR) {
+  if (!isa<VarDecl>(DR->getDecl())) {
+    return true;
+  }
+  const auto *VD = dyn_cast<VarDecl>(DR->getDecl());
```
Should avoid duplicate casts
```cpp
const auto *VD = dyn_cast<VarDecl>(DR->getDecl());
if (!VD) {
    return true;
}
//...
```

ConcreteCactus replied:
Done, I removed all duplicate casts.

vbvictor commented on clang-tools-extra/clang-tidy/bugprone/ConflictingGlobalAccesses.cpp:
```diff
@@ -0,0 +1,748 @@
+}
+
+bool GlobalRWVisitor::isVariable(const Expr *E) {
+  const Type *T = E->getType().getTypePtrOrNull();
+  assert(T);
```
If we get `nullptr` we should not die on assert, just return `false`

vbvictor commented on clang-tools-extra/clang-tidy/bugprone/ConflictingGlobalAccesses.cpp:
```diff
@@ -0,0 +1,748 @@
+
+bool GlobalRWVisitor::handleModifiedVariable(const DeclRefExpr *DR,
+                                             bool IsUnchecked) {
+  if (!isa<VarDecl>(DR->getDecl())) {
```
ditto

ConcreteCactus commented on clang-tools-extra/clang-tidy/bugprone/BugproneTidyModule.cpp:
```diff
@@ -20,6 +20,7 @@
 #include "CastingThroughVoidCheck.h"
 #include "ChainedComparisonCheck.h"
 #include "ComparePointerToMemberVirtualFunctionCheck.h"
+#include "ConflictingGlobalAccesses.h"
```
I renamed the files to end with check, and added the files and changes that the script would add.

ConcreteCactus commented on clang-tools-extra/clang-tidy/bugprone/ConflictingGlobalAccesses.h:
```diff
@@ -0,0 +1,77 @@
+/// \endcode
+///
+/// This option is disabled by default.
+class ConflictingGlobalAccesses : public ClangTidyCheck {
```
Done. I'm guessing LangOpts.CPlusPlus also includes C, as I didn't find a language option that was for just C.

ConcreteCactus commented:
Hi Everyone,
I fixed a crash in the ExecutionVisitor. It wasn't checking whether the definition of a class exists before trying to recurse into the destructor.

Can I get a re-review on this PR? I think I addressed most of the comments.

vbvictor reviewed (commented):
Mostly left comments in docs about general improvements. Didn't look closely at the code, but here are some general recommendations that you should do before major review:

There are a lot of new code that is relatively hard to review, you should try to follow [LLVM coding guidelines](https://llvm.org/docs/CodingStandards.html#style-issues), most importantly these:
- [Don’t Use Braces on Simple Single-Statement Bodies of if/else/loop Statements](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements) (many violations, with this fix the code should become smaller by line count, thus easier to review)
- [Use range-based for loops wherever possible](https://llvm.org/docs/CodingStandards.html#use-range-based-for-loops-wherever-possible)

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/conflicting-global-accesses.rst:
```diff
@@ -0,0 +1,72 @@
+
+Modifying twice or reading and modifying a memory location without a
+defined sequence of the operations is either undefined behavior or has
+unspecified order. This checker is similar to the -Wunsequenced clang warning,
```
```suggestion
unspecified order. This check is similar to the `-Wunsequenced` Clang warning,
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/conflicting-global-accesses.rst:
```diff
@@ -0,0 +1,72 @@
+    int b = (a++) - a; // This is flagged by -Wunsequenced.
+
+However global variables allow for more complex scenarios that
+-Wunsequenced doesn't detect. E.g. ::
```
```suggestion
`-Wunsequenced` doesn't detect. E.g. ::
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/conflicting-global-accesses.rst:
```diff
@@ -0,0 +1,72 @@
+      return globalVar + incFun(); // This is not detected by -Wunsequenced.
+    }
+
+This checker attempts to detect such cases. It recurses into functions that are
```
```suggestion
This check attempts to detect such cases. It recurses into functions that are
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/conflicting-global-accesses.rst:
```diff
@@ -0,0 +1,72 @@
+
+This checker attempts to detect such cases. It recurses into functions that are
+inside the same translation unit. It also attempts not to flag cases that are
+already covered by -Wunsequenced. Global unions and structs are also handled.
```
```suggestion
already covered by `-Wunsequenced`. Global unions and structs are also handled.
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/conflicting-global-accesses.rst:
```diff
@@ -0,0 +1,72 @@
+.. option:: HandleMutableFunctionParametersAsWrites
+
+    It's possible to enable handling mutable reference and pointer function
+    parameters as writes using the HandleMutableFunctionParametersAsWrites
```
```suggestion
    parameters as writes using the `HandleMutableFunctionParametersAsWrites`
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/conflicting-global-accesses.rst:
```diff
@@ -0,0 +1,72 @@
+bugprone-conflicting-global-accesses
+====================================
+
+Finds conflicting accesses on global variables.
```
Please synchronize first statement with Release Notes.

ConcreteCactus commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/conflicting-global-accesses.rst:
```diff
@@ -0,0 +1,72 @@
+
+    It's possible to enable handling mutable reference and pointer function
+    parameters as writes using the HandleMutableFunctionParametersAsWrites
+    option. For example:
```
I clarified the description and added a more descriptive code example. Hope this helps.

ConcreteCactus commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/conflicting-global-accesses.rst:
```diff
@@ -0,0 +1,72 @@
+defined sequence of the operations is either undefined behavior or has
+unspecified order. This checker is similar to the -Wunsequenced clang warning,
+however it only looks at global variables and therefore can find conflicting
+actions recursively inside functions as well.
```
Right, I wasn't sure whether to include/exclude `-Wunsequenced`. I removed the code parts that excluded flagging those cases, and noted later in the documentation that there could be some overlap.

vbvictor commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/conflicting-global-accesses.rst:
```diff
@@ -0,0 +1,72 @@
+    parameters as writes using the HandleMutableFunctionParametersAsWrites
+    option. For example:
+
+    void func(int& a);
```
Also, code-blocks should start with a new line, I'm not sure if they are rendered correctly.
Did you run `ninja docs-clang-tools-html` to check?

vbvictor replied:
You need Sphinx installed in your system.
Please test your docs that they look correct.
https://clang.llvm.org/extra/clang-tidy/Contributing.html#documenting-your-check

ConcreteCactus replied:
Thank you I'll test and fix the documentation and the formatting issues. I'll run clang-tidy as well.

vbvictor commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/unsequenced-global-accesses.rst:
```diff
@@ -0,0 +1,86 @@
+
+.. option:: HandleMutableFunctionParametersAsWrites
+
+When `true`, treat function calls with mutable reference or pointer parameters
```
Whole description of the option should be with 2 spaces ident

vbvictor commented:
Looking at the check code in general, I have some concerns whether this check should be implemented as a Clang Static Analyzer (CSA) check instead of a clang-tidy check. 
My main reason is that we now have a lot of manual node traversing, even an `ExecutionVisitor`. This whole _traversing part_ must be already implemented in the core engine of the CSA. Even in [Getting Involved](https://clang.llvm.org/extra/clang-tidy/Contributing.html#choosing-the-right-place-for-your-check) guide it mentions that CSA should be used for control flow analysis, and we clearly have it here.

I kindly ask for an opinion of clang-tidy maintainers on this matter, @HerrCai0907 @carlosgalvezp @PiotrZSL.

Also, kindly ask CSA maintainers to take a brief look at this checker. From your point of view, will it be easier and less code consuming to implement such checker as a part of CSA? @steakhal @NagyDonat @haoNoQ.

steakhal commented:
CC @michael-jabbour-sonarsource

HerrCai0907 commented:
It does not look like traditional clang-tidy check. Personally I prefer to move it to CSA instead of clang-tidy.

ConcreteCactus commented:
Hi Everyone,

Thank you for the detailed suggestions. Now that I have a better idea of what CSA is, I also think that this check could be better implemented in clang-static-analyzer. Being able to reduce the number of false positives using path sensitive analysis is in my opinion the most important advantage of using CSA.

What do you think? Should I close this PR, and open a new one when I'm ready with the CSA checker?

By the way, I don't have any users. This check is more like a University project I picked. The reason I started with clang-tidy is simply because that was the suggestion my teacher gave me.
